File: src/Auth/Crypto/Password.hs
module Auth.Crypto.Password
  ( PasswordParams (..)
  , StoredPassword (..)
  , hashNewPassword
  , verifyPassword
  ) where

import           Crypto.Hash.Algorithms (SHA256 (..))
import           Crypto.KDF.PBKDF2      (Parameters (..), generate, prfHMAC)
import           Crypto.MAC.HMAC        (HMAC, hmac)
import           Crypto.Random          (getRandomBytes)
import           Data.ByteArray         (constEq, convert)
import qualified Data.ByteString        as BS
import           Data.Text              (Text)
import           Data.Text.Encoding     (encodeUtf8)

data PasswordParams = PasswordParams
  { ppPepper     :: BS.ByteString
  , ppIterations :: Int
  }

data StoredPassword = StoredPassword
  { spSalt :: BS.ByteString
  , spHash :: BS.ByteString
  } deriving (Show, Eq)

saltLength, keyLength :: Int
saltLength = 16
keyLength  = 32

peppered :: BS.ByteString -> Text -> BS.ByteString
peppered pepper password =
  convert (hmac pepper (encodeUtf8 password) :: HMAC SHA256)

derive :: PasswordParams -> BS.ByteString -> BS.ByteString -> BS.ByteString
derive PasswordParams{..} salt pwInput =
  generate (prfHMAC SHA256) (Parameters ppIterations keyLength) pwInput salt

hashNewPassword :: PasswordParams -> Text -> IO StoredPassword
hashNewPassword params password = do
  salt <- getRandomBytes saltLength
  let pwInput = peppered (ppPepper params) password
  pure StoredPassword { spSalt = salt, spHash = derive params salt pwInput }

verifyPassword :: PasswordParams -> Text -> StoredPassword -> Bool
verifyPassword params password StoredPassword{..} =
  let pwInput   = peppered (ppPepper params) password
      candidate = derive params spSalt pwInput
  in constEq candidate spHash

File: src/api/auth/crypto/password.rs
use hmac::{Hmac, Mac};
use rand::RngCore;
use sha2::Sha256;
use subtle::ConstantTimeEq;

const SALT_LENGTH: usize = 16;
const KEY_LENGTH: usize = 32;

#[derive(Debug, Clone)]
pub struct PasswordParams {
    pub pepper: Vec<u8>,
    pub iterations: u32,
}


#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StoredPassword {
    pub salt: Vec<u8>,
    pub hash: Vec<u8>,
}

fn peppered(pepper: &[u8], password: &str) -> Vec<u8> {
    let mut mac = Hmac::<Sha256>::new_from_slice(pepper)
        .expect("HMAC accepts keys of any length");
    mac.update(password.as_bytes());
    mac.finalize().into_bytes().to_vec()
}

fn derive(params: &PasswordParams, salt: &[u8], pw_input: &[u8]) -> Vec<u8> {
    let mut out = vec![0u8; KEY_LENGTH];
    pbkdf2::pbkdf2_hmac::<Sha256>(pw_input, salt, params.iterations, &mut out);
    out
}

pub fn hash_new_password(params: &PasswordParams, password: &str) -> StoredPassword {
    let mut salt = vec![0u8; SALT_LENGTH];
    rand::rng().fill_bytes(&mut salt);
    let pw_input = peppered(&params.pepper, password);
    let hash = derive(params, &salt, &pw_input);
    StoredPassword { salt, hash }
}

pub fn verify_password(params: &PasswordParams, password: &str, stored: &StoredPassword) -> bool {
    let pw_input = peppered(&params.pepper, password);
    let candidate = derive(params, &stored.salt, &pw_input);
    candidate.ct_eq(&stored.hash).into()
}
